actioncable/lib/action_cable/subscription_adapter/redis.rb

```diff
@@ -169,6 +169,10 @@ def ensure_listener_running
                   if retry_connecting?
                     when_connected { resubscribe }
                     retry
+                  else
+                    ActionCable.server.logger.info("Redis reconnect_attempts limit exceeded; restarting.")
+                    @thread = @raw_client = nil
+                    ActionCable.server.restart
```

Comment:
We shouldn't use singleton here (there can be multiple servers, e.g., managed by third-party gems). Every subscription adapter keeps a reference to _its_ server in the `@server` instance variable (`server` reader is available). We should use it here.

Also, no need to reset `@thread`, `@raw_client` instance variables—the whole pubsub object is re-created on restart.

Reply:
> server reader is available). We should use it here.

👍 I think we can change this to `@adapter.server.restart`

> Also, no need to reset @thread, @raw_client instance variables—the whole pubsub object is re-created on restart.

I think you are correct about `@raw_client`. But we do need to reset `@thread` because `shutdown` tries to unsubscribe if `@thread` is not nil, and that won't succeed if Redis is down.